dialogue-test-common/src/main/java/com/palantir/dialogue/AbstractHandshakeTimeoutTest.java

```diff
@@ -0,0 +1,203 @@
+
+    private static final Request request = Request.builder()
+            .body(new RequestBody() {
+                @Override
+                public void writeTo(OutputStream output) throws IOException {
+                    output.write("Hello, World".getBytes(StandardCharsets.UTF_8));
+                }
+
+                @Override
+                public String contentType() {
+                    return "text/plain";
+                }
+
+                @Override
+                public boolean repeatable() {
+                    return true;
+                }
+
+                @Override
+                public void close() {}
+            })
+            .build();
+
+    protected abstract Channel create(ClientConfiguration config);
+
+    private static int getPort(Undertow undertow) {
+        return ((InetSocketAddress)
+                        Iterables.getOnlyElement(undertow.getListenerInfo()).getAddress())
+                .getPort();
+    }
+
+    @Test
+    public void testHandshakeTimeout() throws Exception {
+        SSLContext sslContext = SslSocketFactories.createSslContext(TestConfigurations.SSL_CONFIG);
+        Xnio xnio = Xnio.getInstance(Undertow.class.getClassLoader());
+        DelayingNextTaskExecutorService executor = new DelayingNextTaskExecutorService(Executors.newCachedThreadPool());
+        XnioWorker worker = xnio.createWorkerBuilder()
+                .setWorkerIoThreads(1)
+                .setExternalExecutorService(executor)
+                .build();
+        Undertow server = Undertow.builder()
+                .setWorker(worker)
+                .setHandler(ResponseCodeHandler.HANDLE_200)
+                .addHttpsListener(0, null, sslContext)
+                .build();
+        server.start();
+        try {
+            int serverPort = getPort(server);
+            ClientConfiguration noRetryConfig = ClientConfiguration.builder()
+                    .from(TestConfigurations.create("https://localhost:" + serverPort))
+                    .connectTimeout(Duration.ofMillis(500))
+                    .readTimeout(Duration.ofMillis(500))
+                    .writeTimeout(Duration.ofMillis(500))
+                    .maxNumRetries(0)
+                    .build();
+            Channel noRetryChannel = create(noRetryConfig);
+
+            executor.delayNextTask(Duration.ofSeconds(1));
+
+            assertThatThrownBy(noRetryChannel.execute(TestEndpoint.POST, request)::get)
+                    .getCause()
+                    .satisfies(cause -> assertThat(cause)
+                            .isInstanceOf(SafeLoggable.class)
+                            .isInstanceOf(IOException.class)
+                            .isNotInstanceOf(SocketTimeoutException.class)
+                            .extracting(value -> value.getClass().getSimpleName())
+                            .isEqualTo("SafeConnectTimeoutException"));
+
+            ClientConfiguration retryConfig = ClientConfiguration.builder()
```

Comment:
mind pulling this out into a separate test case so that we have one that verifies the exception that is thrown and another which verifies the retry behaviour? I think having them be named test cases will make it clearer for future readers

Reply:
:+1: